### PR TITLE
Add tests for zero-length scans in scan and associative scan functions (upstream alignment)

### LINKED ISSUE #4781: [split from #4773] functorch/test_control_flow_xpu.py out of sync with upstream scan / associative_scan changes
### 🐛 Describe the bug

parent_issue: 4773
test_cases:
  - op_ut,third_party.torch-xpu-ops.test.xpu.functorch.test_control_flow_xpu.TestControlFlow,test_scan_init_scanned_0
  - op_ut,third_party.torch-xpu-ops.test.xpu.functorch.test_control_flow_xpu.TestControlFlow,test_scan_simple_graph
  - op_ut,third_party.torch-xpu-ops.test.xpu.functorch.test_control_flow_xpu.AssociativeScanTests,test_associative_scan_different_input_size_wrong_dim
---

## Parent Issue

Split from #4773 — [Bug Skip]: [Regression]UT failures 2026-8-2

## Affected Test Cases

- `test/xpu/functorch/test_control_flow_xpu.py::TestControlFlow::test_scan_init_scanned_0`
- `test/xpu/functorch/test_control_flow_xpu.py::TestControlFlow::test_scan_simple_graph`
- `test/xpu/functorch/test_control_flow_xpu.py::AssociativeScanTests::test_associative_scan_different_input_size_wrong_dim`

## Root Cause

`test/xpu/functorch/test_control_flow_xpu.py` is a fork of upstream `pytorch/test/functorch/test_control_flow.py`. Upstream changed the `scan` / `associative_scan` higher-order-op error messages and the traced FX graph, and updated (or deleted) the corresponding tests. The XPU fork was not re-synced, so it asserts against the old expectations. All three failures are pure test-expectation staleness, not XPU kernel bugs.

Per-case detail:

1. **`test_associative_scan_different_input_size_wrong_dim`** — the XPU fork expects `assertRaisesRegex(..., "All xs leaves must at least have.*")`. `torch/_higher_order_ops/associative_scan.py:231` now raises `"All xs leaves must have at least 'dim + 1' dimensions"` (word order changed). Upstream `test/functorch/test_control_flow.py:5950` was already updated to the new regex; the XPU copy was not.

2. **`test_scan_init_scanned_0`** — the XPU fork expects the same stale `"All xs leaves must at least have.*"` regex, but the code now fails earlier with `UncapturedHigherOrderOpError: Expected init and carry to have same metadata`. This test **no longer exists upstream** (removed from `test/functorch/test_control_flow.py`); the XPU fork still carries it. 

3. **`test_scan_simple_graph`** — `assertExpectedInline` mismatch. Upstream's expected graph now numbers the symbolic size nodes `sym_size_int, sym_size_int_1, sym_size_int_2, sym_size_int_3` (see `test/functorch/test_control_flow.py:4567`), while the XPU fork still expects `sym_size_int_1 ... sym_size_int_4`.

**Fix**: re-sync these three tests in `test/xpu/functorch/test_control_flow_xpu.py` with upstream `test/functorch/test_control_flow.py` — update the two `assertRaisesRegex` patterns, update the `assertExpectedInline` graph text, and drop `test_scan_init_scanned_0` (deleted upstream).

## Reproducer

For each affected test case, run individually:

```bash
source ~/.bashrc && cd third_party/torch-xpu-ops/test/xpu && python -m pytest -sxv functorch/test_control_flow_xpu.py -k "test_scan_init_scanned_0"
source ~/.bashrc && cd third_party/torch-xpu-ops/test/xpu && python -m pytest -sxv functorch/test_control_flow_xpu.py -k "test_scan_simple_graph"
source ~/.bashrc && cd third_party/torch-xpu-ops/test/xpu && python -m pytest -sxv functorch/test_control_flow_xpu.py -k "test_associative_scan_different_input_size_wrong_dim"
```

## Representative Error

```
=================================== FAILURES ===================================
___________________ TestControlFlow.test_scan_init_scanned_0 ___________________
torch._dynamo.exc.UncapturedHigherOrderOpError: Expected init and carry to have same metadata but found:
pair[0] differ in 'shape: torch.Size([3, 1, 2]) vs torch.Size([3, 3, 2])', 'stride: (2, 2, 1) vs (6, 2, 1)', where lhs is FakeTensor(..., size=(3, 1, 2)) and rhs is FakeTensor(..., size=(3, 3, 2))
  Higher Order Operator: torch.ops.higher_order.scan

from user code:
   File "/home/tszulist/src/pytorch/torch/_higher_order_ops/scan.py", line 215, in run_flattened_scan
    return scan_op(combine_fn, leaves_init, leaves_xs, ())

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File ".../test/xpu/functorch/test_control_flow_xpu.py", line 2337, in test_scan_init_scanned_0
    with self.assertRaisesRegex(
  File "/usr/lib/python3.12/unittest/case.py", line 276, in __exit__
    self._raiseFailure('"{}" does not match "{}"'.format(
  File "/usr/lib/python3.12/unittest/case.py", line 200, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: "All xs leaves must at least have.*" does not match "Expected init and carry to have same metadata but found: ..."
```

Second signature (`test_associative_scan_different_input_size_wrong_dim`):

```
ValueError: All xs leaves must have at least 'dim + 1' dimensions

During handling of the above exception, another exception occurred:

  File ".../test/xpu/functorch/test_control_flow_xpu.py", line 4853, in test_associative_scan_different_input_size_wrong_dim
    with self.assertRaisesRegex(
AssertionError: "All xs leaves must at least have.*" does not match "All xs leaves must have at least 'dim + 1' dimensions"
```

Third signature (`test_scan_simple_graph`):

```
AssertionError: 'def [99 chars]e_int_1 = torch.ops.aten.sym_size.int(init_1, [674 chars]p_1)' != 'def [99 chars]e_int = torch.ops.aten.sym_size.int(init_1, 1)[668 chars]p_1)'
+     sym_size_int = torch.ops.aten.sym_size.int(init_1, 1)
-     sym_size_int_1 = torch.ops.aten.sym_size.int(init_1, 1)
...
: To accept the new output, re-run test with envvar EXPECTTEST_ACCEPT=1
```

### Versions

- Latest good PyTorch: [f2fa815](https://github.com/pytorch/pytorch/commit/f2fa81593a74bb8aca2c3da562765378665af089)
- Current (failing) PyTorch: [4ee8fea](https://github.com/pytorch/pytorch/commit/4ee8fea1f8cbb27acf0dac1c4f2e8b8bb6aef7f7)
- Platform: Intel(R) Data Center GPU Max 1100, Ubuntu 24.04, libze1 1.24.0.0, intel-opencl-icd 25.18.33578.51
- XPU build: 20260100
- Reproduced locally on `torch 2.14.0a0+gitfdd065a`, torch-xpu-ops `1e6e7fbe`.

### PR DESCRIPTION
**Pull request by tszulist-hbn**
Resolves https://github.com/intel/torch-xpu-ops/issues/4781

Align `test_control_flow_xpu.py` with upstream PyTorch after pytorch/pytorch#188348 ("[HOP] Fixes for zero-length scan dim"), which made a zero-length scan dimension valid for `scan` and `associative_scan` and changed two error messages.

Mirrors the test changes from that commit:

- `test_scan_init_scanned_0` is removed. It asserted that a zero-length scan dim raises `"All xs leaves must at least have.*"`; that case is now supported, so there is no error left to assert. Upstream deleted it for the same reason. 
- New `test_scan_zero_length`, `test_scan_zero_length_autograd` and `test_associative_scan_zero_length` cover the now-supported behavior, and `test_scan_unequal_scan_dim_size` keeps error-path coverage for the check that is still an error (mismatched scan dim sizes across `xs` leaves).
- `test_scan_simple_graph`: updated the `sym_size_int*` numbering in the golden graph string.
- `test_associative_scan_different_input_size_wrong_dim`: updated the regex to the new exact message, `"All xs leaves must have at least 'dim + 1' dimensions"`.
- Dropped the commented-out `decorateIf` block that upstream removed from `test_associative_scan_compile`.

The new `test_associative_scan_zero_length` uses this file's `DEVICE_TYPE` / `requires_accelerator` conventions instead of upstream's CUDA-specific decorators.

### CODE REVIEW

File: test/xpu/functorch/test_control_flow_xpu.py [hunk 1]
```diff
@@ -2325,25 +2325,63 @@ def test_scan_compile_cnt(self, reverse, device):
...
 
     @skipIfTorchDynamo("don't test compile on compile")
-    def test_scan_init_scanned_0(self):
-        # Only init and no input
-        x = torch.randn(3, 1, 2, device=torch.device("cpu"))
-        init = torch.randn(3, 2, device=torch.device("cpu"))
-        dim = 1
+    @parametrize("reverse", [False, True])
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
+    def test_scan_zero_length(self, reverse, compile_mode):
```
[the remark this answers is not in the document]
Comment (tszulist-hbn) [reply test/xpu/functorch/test_control_flow_xpu.py:2330]: Invalid - the decorator is already there.

**Review by AKloniecki — APPROVED** [review]
LGTM

**Review by pbielak — APPROVED** [review]

**Comment by tszulist-hbn** [comment]
Representative matching failures occur in PR #4818, PR #4897 and PR #4888.
@torchxpubot merge -f "Random/unrelated to the PR CI failures"

**Comment by tszulist-hbn** [comment]
@torchxpubot merge -f "Random/unrelated to the PR CI failures"
